### DISCUSSION TITLE
Form Layout Cards

### DISCUSSION THREAD

**Discussion by johncarter-** [filamentphp/filament#112] [Ideas]
I would be interested to see if there would be a way to introduce a "Card" layout that could allow us to wrap separate sections of the form in card components?
#### Before
![image](https://user-images.githubusercontent.com/3776888/110237671-10654a80-7f35-11eb-8376-44c72593b75f.png)

#### After 
![image](https://user-images.githubusercontent.com/3776888/110238138-c03bb780-7f37-11eb-85e9-0943f215c432.png)

I am not sure how best this could be implemented, or is it already possible?

**Comment by johncarter-** [comment]
I'll glady have a crack a PR for this if you like the idea.

**Reply by ryangjchandler** [reply]
@johncarter- Have a go at a PR and see if you can find a nice API for this. I'm quite interested in the idea!

**Reply by danharrin** [reply]
At the moment, `schema()` only accepts an array. We should...
- Check if the first param is an array. If it is an array, wrap the contents in a card component (to achieve the current behaviour).
- If the first param is a form component, render it without a card wrapper. It could be a `Grid` which can have a couple of cards in its own schema to achieve this layout.

**Reply by ryangjchandler** [reply]
@danharrin How about a new `Card` layout component though or even a `card` modifier for the `Grid` component..

**Reply by danharrin** [reply]
```php
Form::make()->schema(
    Grid::make([
        Card::make([
            // Main card schema
        ])->columnSpan(3),
        Card::make([
            // Sidebar card schema
        ]),
    ])->columns(4),
);
```

**Reply by ryangjchandler** [reply]
Yeah, exactly that. I think in this case the `columnSpan()` should only be 2, since it's 3 columns wide but that's what I had in mind yeah.

**Reply by danharrin** [reply]
Looks like 4 to me 🧐 I need my eyesight tested.

**Reply by johncarter-** [reply]
I thought about the card component, but how would you remove the card that exists on all forms in that case?

I was thinking maybe something chainable, like `asCard(false)`?

**Reply by ryangjchandler** [reply]
The other thing to consider here is the placement of the "Create" button too. Should it always be outside of a card / grid, maybe you want to show it inside of the sidebar card instead.

**Reply by danharrin** [reply]
> I thought about the card component, but how would you remove the card that exists on all forms in that case?
> 
> I was thinking maybe something chainable, like `asCard(false)`?

The card would only be a wrapper if the first parameter of `schema()` is an array. Else, there is no wrapper. No need for an extra method :)

**Reply by danharrin** [reply]
> The other thing to consider here is the placement of the "Create" button too. Should it always be outside of a card / grid, maybe you want to show it inside of the sidebar card instead.

Maybe we could have an `Actions` component that you can choose where to render.

**Reply by ryangjchandler** [reply]
Yeah, that could work. Then it would just be a case of checking if an actions component is registered anywhere and hiding the default actions at the bottom?

**Reply by danharrin** [reply]
> Yeah, that could work. Then it would just be a case of checking if an actions component is registered anywhere and hiding the default actions at the bottom?

Nah, we could just render the actions in the default card if it gets added automatically. It's just a case of moving the form slot about.

**Reply by ryangjchandler** [reply]
Oh yeah, duh.

**Reply by johncarter-** [reply]
I'll can probably make a start on this today unless you guys want to take it on?

**Reply by danharrin** [reply]
> I'll can probably make a start on this today unless you guys want to take it on?

I don't mind either way!

**Reply by johncarter-** [reply]
I have taken a look at it this morning and I have to say that there are too many moving parts for me to wrap my bonce round.

I admit defeat and I'll leave it to you guys.

**Reply by danharrin** [reply]
No worries @johncarter-! We'll work on a solution.

**Comment by danharrin** [comment]
This feature has been written and will be present in the next minor release (v1.8.0).

**Reply by topegret** [reply]
try to find similar solution. now it is 2.x , how to make this happen. from what I read, card are full length and i am not able to put card into grids. or did I miss something? thanks

**Reply by danharrin** [reply]
Hey! Check the demo repo, we have examples there of this layout

**Reply by topegret** [reply]
sure thanks. was not aware there is a demo /demo repo.  will sure check it out.
